子コンポーネントの入力を受け取り、APIで住所をfetchして描画する機能の実装

WATA-Haru opened the pull request:
https://github.com/WATA-Haru/reviewMe/issues/21

変更差分が多いですがほとんど`A15/public/mockServiceWorker.js`です。
自動生成のファイルなので見なくて大丈夫です。

## 実装
### 1. APIリクエストをする際に実装ミスがあったので下記で修正しました。
APIエンドポイントのミス
  - [x] APIエンドポイント
は`https://zipcloud.ibsnet.co.jp/api/search`が正しいのですが、実装で末尾に`/`があったので削除しました。
  - [x] [FIX] APIエンドポイントの修正(末尾スラッシュの削除) [9535272](https://github.com/WATA-Haru/reviewMe/pull/24/commits/95352724e1a5524da5daee4fb3457e0dfda4def2)

### 2. `new URL`の挙動が予想と違ったので単純な連結に変更しました。
  - [ ] `zipCloudURLWrapper`を`new URL`で実装していましたが、ほしい機能が単純なURL同士の連結だったので、単純な文字列結合に変更しました。
  - [ ] [FIX] zipCloudURLWrapperをnew URLではなくURLを連結するように修正 [3991d2b](https://github.com/WATA-Haru/reviewMe/pull/24/commits/3991d2b44323569c1107028f725c0c05c61954f2)

### 3. APIモックをブラウザでも使えるように設定しました
こちらで対応しました。
[ADD] browserのAPIモックを作成
[67699a9](https://github.com/WATA-Haru/reviewMe/pull/24/commits/67699a975f28fd3a7a102f29d7260655659ca949)

### 4. 子コンポーネントの入力を受け取り、APIで住所をfetchし、描画する機能の仮実装しました
- [ADD] 仮実装 入力値を親コンポーネントに受け渡すように追加
[efe178e](https://github.com/WATA-Haru/reviewMe/pull/24/commits/efe178e86af8a3ddeb9baf365d68309c90643697)

## review観点

- 入力されたデータを`emit`で受け渡して親コンポーネントで描画するという書き方があっているのか、別の方法がないか疑問なので、よりよい実装がないか見ていただきたいです。(今は住所をcontainerで直接描画していますが、今後は`props`で渡して描画します。)
[ADD] 仮実装 入力値を親コンポーネントに受け渡すように追加
[efe178e](https://github.com/WATA-Haru/reviewMe/pull/24/commits/efe178e86af8a3ddeb9baf365d68309c90643697)




https://github.com/user-attachments/assets/13c14325-3834-41d3-8c7f-755c3476c97c

WATA-Haru committed:
[FIX] APIエンドポイントの修正(末尾スラッシュの削除)

末尾にスラッシュが含まれていたが、エンドポイントにはなかったので削除した
`https://zipcloud.ibsnet.co.jp/api/search`
[document](https://zipcloud.ibsnet.co.jp/doc/api)

WATA-Haru committed:
[FIX] zipCloudURLWrapperをnew URLではなくURLを連結するように修正

[newURL](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL)の挙動をURLを連結するものだと誤認していたので、単純にURLに文字列を連結する関数に修正

WATA-Haru committed:
[ADD] browserのAPIモックを作成

node環境ではなくブラウザでMSWが使用できるよう変更

WATA-Haru committed:
[ADD] 仮実装 入力値を親コンポーネントに受け渡すように追加

hanzzoo commented on A15/.env:
```diff
@@ -1 +1 @@
-VITE_ZIP_API_URL = 'https://zipcloud.ibsnet.co.jp/api/search/'
+VITE_ZIP_API_URL = 'https://zipcloud.ibsnet.co.jp/api/search'
```
[nits]
`.env` ファイルは基本的にコミットせず、`.gitignore`対象になる。コンフィデンシャルな値を扱うこともあるので、
`.env.sample` のようなファイルを用意したりすることが結構あるね

環境変数の名前だけ書いてたりする

```
VITE_ZIP_API_URL = ''
```

WATA-Haru replied:
@hanzzoo 
ありがとうございます。対応しました！
[[ADD] .env.sampleに空のAPIエンドポイントを追加](https://github.com/WATA-Haru/reviewMe/pull/24/commits/62342748048b52dd45f554e73d9fd52d6533c96b)

hanzzoo commented on A15/src/main.ts:
```diff
@@ -3,4 +3,21 @@ import './assets/main.css'
+ * @description - setup msw on browser env official example(https://github.com/mswjs/examples/blob/main/examples/with-vue/src/main.ts)
+ */
+async function prepareApp() {
+  if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
```
[nits]
Viteでは `meta.env.VITE_XXX` という指定の仕方の方が推奨されている。

hanzzoo replied:
今のままだと、本番とDevでAPIのエンドポイントを切り替える術がないので、`VITE_ENV` とかで切り替える仕組みを作った方が良さそう！

WATA-Haru replied:
@hanzzoo 

>Viteでは meta.env.VITE_XXX という指定の仕方の方が推奨されている。

ありがとうございます。NODEのものと混ざっていたので修正しました。勉強になりました。

[[FIX] import.meta.env.DEVでmockサーバを切り替えできるように設定](https://github.com/WATA-Haru/reviewMe/pull/24/commits/50fe54f92fcc8e38ba07dc5e321fef385cf4983a)
[50fe54f](https://github.com/WATA-Haru/reviewMe/pull/24/commits/50fe54f92fcc8e38ba07dc5e321fef385cf4983a)


>今のままだと、本番とDevでAPIのエンドポイントを切り替える術がないので、VITE_ENV とかで切り替える仕組みを作った方が良さそう！

こちらは`.env.development`と`env.production`を作成しました。(git登録は`.env.development.sample`と`.env.production.sample`でしています)
https://ja.vite.dev/guide/env-and-mode.html#env-files

[[ADD] developmentとproductionでAPIエンドポイントを変えられるようにviteで設定](https://github.com/WATA-Haru/reviewMe/pull/24/commits/e9b5de0b410495b00d16dc9b87ee78102ca36a76)
[e9b5de0](https://github.com/WATA-Haru/reviewMe/pull/24/commits/e9b5de0b410495b00d16dc9b87ee78102ca36a76)

hanzzoo commented on A15/src/components/Form/container/index.vue:
```diff
@@ -1,10 +1,48 @@
+
+    if (!response) {
+      throw new Error('fetch failed')
+    }
```
[nits]
`A15/src/components/ZipCodeInput/container/index.vue` で タイプガードされているので、`null` が入ってくることは無さそう！

```ts
  const zipcode = createZipCode(textRefFirst.value.concat(textRefSecond.value))
  if (zipcode) {
    emits('send-zipcode-from-zip-code-input', zipcode)
  }
```

WATA-Haru replied:
@hanzzoo 
こんなところまで！！気づいていただいてありがとうございます。
確かに`zipcode`を`emit`する際にタイプガードしているのでnullは入りませんね。ありがとうございます！

[[FIX] nullが入らない部分の余分なタイプガードの削除](https://github.com/WATA-Haru/reviewMe/pull/24/commits/35ea144dfe344a1ba4fe44ba4c2e92cc44ffe146)
[35ea144](https://github.com/WATA-Haru/reviewMe/pull/24/commits/35ea144dfe344a1ba4fe44ba4c2e92cc44ffe146)

hanzzoo reviewed (commented):
コメントしました！

WATA-Haru committed:
[DELETE] `.env`の削除と`.env.develop`と`.env.production`を`.gitignore`に追加

WATA-Haru committed:
[ADD] `.env.sample`に空のAPIエンドポイントを追加

WATA-Haru committed:
[ADD] developmentとproductionでAPIエンドポイントを変えられるようにviteで設定

WATA-Haru committed:
[FIX] testモードでもdevの環境変数が使えるように`vitest --mode developmentをデフォルトに設定

WATA-Haru committed:
[FIX] import.meta.env.DEVでmockサーバを切り替えできるように設定

WATA-Haru committed:
[FIX] APIエンドポイントの直書きをimport.meta.env.VITE_ZIP_API_URLに変更

WATA-Haru committed:
[FIX] nullが入らない部分の余分なタイプガードの削除

darienjonathan commented on A15/.env.development.sample:
```diff
@@ -0,0 +1 @@
+VITE_ZIP_API_URL = ''
```
[imo]

最後の行に改行を入れておきましょう。改行を自動に入れてくれる設定は各IDEにあると思います。

改行を入れる理由は、[POSIXの定義](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206)によると、行とは、改行で終端される、0個以上の改行以外の文字列であることです。

WATA-Haru replied:
ありがとうございます。改行を追加するルールをeditorに着けましたので今後は防止できます！


[[FIX] envファイルに末尾改行を追加](https://github.com/WATA-Haru/reviewMe/pull/24/commits/06cd774995f023d8fc8061fb2b4f2b3bcc4c31c7)


念のためlintにもルールを追加してみました！
[[FIX] eslintで末尾改行がない場合にエラーを表示するルール追加](https://github.com/WATA-Haru/reviewMe/pull/24/commits/1998b3ed031dcc9a830246ffcefbe915a8dcb112)

こちらです。
https://github.com/WATA-Haru/reviewMe/pull/24/commits/1998b3ed031dcc9a830246ffcefbe915a8dcb112#diff-ad10049806628484266a0de55cf604381772acf7eae24a6d36de065b438141f0R25-R29

darienjonathan commented on A15/src/utils/zipCloudURLWrapper/index.ts:
```diff
@@ -4,5 +4,6 @@ export const zipCloudURLWrapper = (path: string | null = null): string => {
     return new URL(base).toString()
   }
-  return new URL(path, base).toString()
+
+  return new URL(base).toString() + path
```
[imo]

これって、クエリ文字列を対応するための変更であっていますか？もしそうであれば、クエリ文字列専用のパラメーターを用意し、`URL.searchParams` で対応したほうがいいのではないかと思います。

https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams

WATA-Haru replied:
@darienjonathan 

呼び出し側が`zipCloudURLWrapper()`を呼び出すと、`base`に設定されている`https://zipcloud.ibsnet.co.jp/api/search`が返ってきて、パラメータを設定すると`https://zipcloud.ibsnet.co.jp/api/search?params=hoge`のようなurlが返ってくる関数を作成したかったです。
また、`URL.searchParams`を知らなかったのでご教示いただきありがとうございます。


**修正案**
私はこの関数があまり適切でないと思っています。
そのため、関数自体を以下のように、オブジェクトでパラメータを渡して、それに対応するパラメータをURLに着ける形に修正しようと思ったのですがいかがでしょうか？
また、今回のPRの範囲とは少し違う気がしたので別PRで対応したいと思ったのですがよろしいでしょうか。


**疑似コード**
```
/**
* @example
* const fullURL = zipCloudURLWrapper({ zipcode: "1111111", limit: 1})
* console.log(fullURL)
* 
* output
* https://zipcloud.ibsnet.co.jp/api/search?zipcode=1111111&limit=1
*/
export const zipCloudURLWrapper = (
{ zipcode: string, 
  callback: string, 
  limit: number
} | null = null): string => {

  const base = import.meta.env.VITE_ZIP_API_URL

  if (path === null || path === '') {
    return new URL(base).toString()
  }

  // TODO
  // 受け取ったparamsを結合する処理
}
```


**補足**
[APIの仕様](https://zipcloud.ibsnet.co.jp/doc/api
)では、取ることのできるパラメータは以下のようになっていました。

|パラメータ名 | 項目名 | 必須 | 備考|
|-- | -- | -- | --|
|zipcode | 郵便番号 | ○ | 7桁の数字。ハイフン付きでも可。完全一致検索。|
|callback | コールバック関数名 | - | JSONPとして出力する際のコールバック関数名。UTF-8でURLエンコードした文字列。|
|limit | 最大件数 | - | 同一の郵便番号で複数件のデータが存在する場合に返される件数の上限値（数字）　※デフォルト：20|

darienjonathan commented on A15/src/utils/fetchAddressFromZipCode/index.ts:
```diff
@@ -4,9 +4,9 @@ import { type AxiosResponse } from 'axios'
-const instance = axios.create({
-  baseURL: zipCloudURLWrapper(''),
-})
+//const instance = axios.create({
+//  baseURL: zipCloudURLWrapper(''),
+//})
```
[ask]

コメントアウトするだけで、消さない理由はなんでしょう？

WATA-Haru replied:
消し忘れました！ありがとうございます。こちらで削除しました
[7648566](https://github.com/WATA-Haru/reviewMe/pull/24/commits/7648566b9592111a862195ecd43caa5950796598)

darienjonathan commented on A15/src/components/Form/container/index.vue:
```diff
@@ -1,10 +1,45 @@
   name: 'FormContainer',
 })
+
+const addressRef = ref({
+  address1: '',
+  address2: '',
+  address3: '',
+})
```
[imo]
レスポンスも同じ形になっているので仕方ないかもしれませんが、やはり 数字 (1, 2, 3) で分けると違いがわからないなぁと思います。

例えば、
- `addressPrefecture`
- `addressCity` （区町村は `City` じゃないため違和感ある場合は `addressMunicipal` でも良さそうかもしれません）
- `addressBlock` / `addressNeighborhood`

とかはどうでしょう？でもレスポンスも 1, 2, 3 を使っているのでもしこのままにしたい場合はそれで良くて、コメントだけを残したほうが良いと思います。

WATA-Haru replied:
変数名のご提案までいただきましてありがとうございます。
変数名を変更いたしました。
[3d40716](https://github.com/WATA-Haru/reviewMe/pull/24/commits/3d40716a755c4cc721582e23a252470f5aec6a5b)

よくよく考えたら
今回のaddressRefは受け取ったレスポンスをフロント側で扱う際の変数なので、命名は直感的にわかりやすい変数名のした方がよさそうに思ったので変更に賛成です！

WATA-Haru replied:
>レスポンスも 1, 2, 3 を使っているのでもしこのままにしたい場合はそれで良くて、コメントだけを残したほうが良いと思います。


私もここが心配なので、ダリエンさんのおっしゃっていることとは逆になりますが、addressRefのdescriptionの`@description`に、APIからもらったaddress1,address2, address3が格納されることを追記しました。
いかがでしょうか？

[e7a6b8f](https://github.com/WATA-Haru/reviewMe/pull/24/commits/e7a6b8f730ca2ca4b3291b452737460e6530a702)

darienjonathan commented on A15/src/components/Form/container/index.vue:
```diff
@@ -1,10 +1,45 @@
+    addressRef.value.address1 = result.address1
+    addressRef.value.address2 = result.address2
+    addressRef.value.address3 = result.address3
+  } catch (error) {
+    console.error(error)
+    return
+  }
```
[imo]

Wantの要件に「API 通信が失敗した際は失敗した旨が画面上でわかるようにする（見た目はこだわらない）」があるので、何かの例外対応をしたほうがいいと思います。もし別PRで対応する場合は、TODOコメントを残すのはどうでしょう？

WATA-Haru replied:
ありがとうございます。おっしゃる通りです。こちらでTODOコメント追記しました！

[ebc0599](https://github.com/WATA-Haru/reviewMe/pull/24/commits/ebc0599fdea9d9fdfe7439ef5910a1574b5099c5)

darienjonathan commented on A15/src/components/Form/container/index.vue:
```diff
@@ -1,10 +1,45 @@
+const handleZipCode = async (zipcode: ZipCode) => {
+  try {
+    const response = await fetchAddressFromZipCode(zipcode)
+    const result = response?.data?.results?.[0]
```
[imo]

`0` の配列アクセスについて何かの説明を書いたほうがいいと思います。

例えば、
```suggestion
    // レスポンスの最初の住所だけを使用
    const firstResultIndex = 0
    const result = response?.data?.results?.[firstResultIndex]
```

WATA-Haru replied:
ありがとうございます。こちらコメント追加と変数への変更を行いました。

[849a883](https://github.com/WATA-Haru/reviewMe/pull/24/commits/849a8839df3d2d59df4557d6595ff8eb92d821f6)

darienjonathan commented on A15/.gitignore:
```diff
@@ -28,3 +28,6 @@ coverage
 *.sw?
 
 *.tsbuildinfo
+
```
[must]

```suggestion
.env
```

WATA-Haru replied:
.envを追加し忘れました。こちらで対応しました
[771d603](https://github.com/WATA-Haru/reviewMe/pull/24/commits/771d603a7eaa5edcef00469439a45a9baca7fe10)

WATA-Haru committed:
[FIX] envファイルに末尾改行を追加

WATA-Haru committed:
[FIX] eslintで末尾改行がない場合にエラーを表示するルール追加

WATA-Haru committed:
[FIX] mockServiceWorkerがtslintをignoreするように設定

WATA-Haru committed:
[FIX] 無駄なコメント削除

WATA-Haru committed:
[ADD] TODOコメントをFormのcontainerに追加(API通信時)

WATA-Haru committed:
[FIX] .envを.gitignoreへ追加

WATA-Haru committed:
[FIX] address1, address2, address3の変数名を変更

zipCodeのAPIの戻り値を格納する変数をaddress1, address2, address3としていましたが、直感的でないので
addressPrefecture, addressCity, addressNeighborhood に変換しました。

WATA-Haru committed:
[ADD] addressRefとAPIの戻り値を対応付けのためにjsdocのdescriptionコメントを追加

WATA-Haru committed:
[FIX] indexへの直接のアクセスをfirstResultIndexを使った変数のアクセスに変更
